folding-schemes/src/folding/hypernova/mod.rs

```diff
@@ -25,3 +55,535 @@ impl<F: PrimeField> Witness<F> {
+            F.clone(),
+            pp.ccs.clone(),
+        )?;
+        let ccs = augmented_f_circuit.ccs.clone();
+
+        let cf_circuit = CycleFoldCircuit::<C1, GC1>::empty();
+        let cf_r1cs = get_r1cs_from_cs::<C2::ScalarField>(cf_circuit)?;
+
+        // setup the dummy instances
+        let W_dummy = Witness::<C1::ScalarField>::dummy(&ccs);
+        let U_dummy = LCCCS::<C1>::dummy(ccs.l, ccs.t, ccs.s);
+        let w_dummy = W_dummy.clone();
+        let mut u_dummy = CCCS::<C1>::dummy(ccs.l);
+        let (cf_W_dummy, cf_U_dummy): (NovaWitness<C2>, CommittedInstance<C2>) =
+            cf_r1cs.dummy_instance();
+        u_dummy.x = vec![
+            U_dummy.hash(
+                &pp.poseidon_config,
+                C1::ScalarField::zero(),
+                z_0.clone(),
+                z_0.clone(),
+            )?,
+            cf_U_dummy.hash_cyclefold(&pp.poseidon_config)?,
+        ];
+
+        // W_dummy=W_0 is a 'dummy witness', all zeroes, but with the size corresponding to the
+        // R1CS that we're working with.
+        Ok(Self {
+            _gc1: PhantomData,
+            _c2: PhantomData,
+            _gc2: PhantomData,
+            ccs,
+            cf_r1cs,
+            poseidon_config: pp.poseidon_config.clone(),
+            cs_params: pp.cs_params.clone(),
+            cf_cs_params: pp.cf_cs_params.clone(),
+            F,
+            i: C1::ScalarField::zero(),
+            z_0: z_0.clone(),
+            z_i: z_0,
+            W_i: W_dummy,
+            U_i: U_dummy,
+            w_i: w_dummy,
+            u_i: u_dummy,
+            // cyclefold running instance
+            cf_W_i: cf_W_dummy,
+            cf_U_i: cf_U_dummy,
+        })
+    }
+
+    /// Implements IVC.P of HyperNova+CycleFold
+    fn prove_step(
+        &mut self,
+        mut rng: impl RngCore,
+        external_inputs: Vec<C1::ScalarField>,
+    ) -> Result<(), Error> {
+        let augmented_f_circuit: AugmentedFCircuit<C1, C2, GC2, FC>;
+
+        if self.z_i.len() != self.F.state_len() {
+            return Err(Error::NotSameLength(
+                "z_i.len()".to_string(),
+                self.z_i.len(),
+                "F.state_len()".to_string(),
+                self.F.state_len(),
+            ));
+        }
+        if external_inputs.len() != self.F.external_inputs_len() {
+            return Err(Error::NotSameLength(
+                "F.external_inputs_len()".to_string(),
+                self.F.external_inputs_len(),
+                "external_inputs.len()".to_string(),
+                external_inputs.len(),
+            ));
+        }
+
+        if self.i > C1::ScalarField::from_le_bytes_mod_order(&usize::MAX.to_le_bytes()) {
+            return Err(Error::MaxStep);
+        }
+
+        let mut transcript_p: PoseidonTranscript<C1> =
+            PoseidonTranscript::<C1>::new(&self.poseidon_config.clone());
+        let (nimfs_proof, U_i1, W_i1, rho_bits) = NIMFS::<C1, PoseidonTranscript<C1>>::prove(
```

Comment:
not sure I follow, do you mean to include the NIMFS::prove logic inside the if `i=zero`?

Reply:
No, I mean that if `i=0`, $NIMFS.P$ should not be running right? Hence, I was wondering why $NIMFS.P$ was not located below, where this case is handled.

Reply:
for i=0, NIMFS.P runs in order to obtain the $U_{i+1}$ instance, folding the initial base instances ('dummy'), in order to compute the 'x' (`u_i1_x`) which is used in the AugmentedFCircuit. In Nova for `U_i1` can use a dummy instance (not from doing the actual NIFS.P) because it's params would be the same, but in hypernova found out that the v params depend on it, so ended up using the NIMFS.P for all cases (including i=0)

Reply:
update: you are right, thanks for pointing it out! Updated it so that in the base case (i=0) the nimfs.p is not computed and `None` is used (which the circuit will  allocate a `NIMFSProof::dummy()`), also now the hash check differentiates between the base case and the non-basecase
(changes diff [here](https://github.com/privacy-scaling-explorations/sonobe/compare/b6750b58f759980312d74e1679f3e0bd3d2c7bbe..2813e79c8a700e175798d7fa848d6c4a380f67c1))